File: data/plugins/skill-runecraft/rune.rb
require 'java'

java_import 'org.apollo.game.model.def.ItemDefinition'

# The list of runes.
RUNES = {}

# Represents a rune that can be crafted.
class Rune
  attr_reader :name, :id, :level, :experience
  
  def initialize(id, level, experience, multiplier)
    @id = id
    @name = ItemDefinition.lookup(id).name.downcase
    @level = level
    @experience = experience
    @multiplier = multiplier
  end

  def multiplier(level)
    return @multiplier.call(level)
  end

  def equals(other)
    return (get_class == other.get_class && id == other.id)
  end
end

# Appends a rune to the list.
def append_rune(hash)
  raise 'Hash must contain an id, experience, .' unless hash.has_key?(:id) && hash.has_key?(:level) && hash.has_key?(:experience) && hash.has_key?(:multiplier)
  id = hash[:id]; altar = hash[:altar]; level = hash[:level]; experience = hash[:experience]; multiplier = hash[:multiplier]
  
  RUNES[altar] = Rune.new(id, level, experience, multiplier)
end

append_rune :name => :air_rune,    :altar => 2478, :id => 556, :level => 1,  :experience =>    5, :multiplier => lambda { |level| (level / 11).floor + 1 }
append_rune :name => :mind_rune,   :altar => 2479, :id => 558, :level => 1,  :experience =>  5.5, :multiplier => lambda { |level| (level / 14).floor + 1 }
append_rune :name => :water_rune,  :altar => 2480, :id => 555, :level => 5,  :experience =>    6, :multiplier => lambda { |level| (level / 19).floor + 1 }
append_rune :name => :earth_rune,  :altar => 2481, :id => 557, :level => 9,  :experience =>  6.5, :multiplier => lambda { |level| (level / 26).floor + 1 }
append_rune :name => :fire_rune,   :altar => 2482, :id => 554, :level => 14, :experience =>    7, :multiplier => lambda { |level| (level / 35).floor + 1 }
append_rune :name => :body_rune,   :altar => 2483, :id => 559, :level => 20, :experience =>  7.5, :multiplier => lambda { |level| (level / 46).floor + 1 }
append_rune :name => :cosmic_rune, :altar => 2484, :id => 564, :level => 27, :experience =>    8, :multiplier => lambda { |level| level >= 59 ? 2 : 1    }
append_rune :name => :chaos_rune,  :altar => 2487, :id => 562, :level => 35, :experience =>  8.5, :multiplier => lambda { |level| level >= 74 ? 2 : 1    }
append_rune :name => :nature_rune, :altar => 2486, :id => 561, :level => 44, :experience =>    9, :multiplier => lambda { |level| level >= 91 ? 2 : 1    }
append_rune :name => :law_rune,    :altar => 2485, :id => 563, :level => 54, :experience =>  9.5, :multiplier => lambda { |level| 1                      }
append_rune :name => :death_rune,  :altar => 2488, :id => 560, :level => 65, :experience =>   10, :multiplier => lambda { |level| 1                      }

File: game/plugin/skills/runecrafting/src/Rune.kt
package org.apollo.game.plugin.skill.runecrafting

import org.apollo.game.plugin.skill.runecrafting.Altar.*

interface Rune {
    /**
     * The item id of the rune.
     */
    val id: Int

    /**
     * The [Altar] this rune must be crafted at.
     */
    val altar: Altar

    /**
     * The runecrafting level required to craft runes of this type.
     */
    val level: Int

    /**
     * The amount of experience rewarded from crafting a single rune of this type.
     */
    val xp: Double

    /**
     * Get the multiplier that is applied to the number of runes the player crafts for their runecrafting level.
     *
     * [playerLevel] - The players current runecrafting level.
     */
    fun getBonusMultiplier(playerLevel: Int): Double
}

enum class DefaultRune(
    override val id: Int,
    override val altar: Altar,
    override val level: Int,
    override val xp: Double
) : Rune {
    AIR_RUNE(556, AIR_ALTAR, 1, 5.0),
    MIND_RUNE(558, MIND_ALTAR, 1, 5.5),
    WATER_RUNE(555, WATER_ALTAR, 5, 6.0),
    EARTH_RUNE(557, EARTH_ALTAR, 9, 6.5),
    FIRE_RUNE(554, FIRE_ALTAR, 14, 7.0),
    BODY_RUNE(559, BODY_ALTAR, 20, 7.5),
    COSMIC_RUNE(564, COSMIC_ALTAR, 27, 8.0),
    CHAOS_RUNE(562, CHAOS_ALTAR, 35, 8.5),
    NATURE_RUNE(561, NATURE_ALTAR, 44, 9.0),
    LAW_RUNE(563, LAW_ALTAR, 54, 9.5),
    DEATH_RUNE(560, DEATH_ALTAR, 65, 10.0);

    override fun getBonusMultiplier(playerLevel: Int): Double = when (this) {
        DefaultRune.AIR_RUNE -> (Math.floor((playerLevel / 11.0)) + 1)
        DefaultRune.MIND_RUNE -> (Math.floor((playerLevel / 14.0)) + 1)
        DefaultRune.WATER_RUNE -> (Math.floor((playerLevel / 19.0)) + 1)
        DefaultRune.EARTH_RUNE -> (Math.floor((playerLevel / 26.0)) + 1)
        DefaultRune.FIRE_RUNE -> (Math.floor((playerLevel / 35.0)) + 1)
        DefaultRune.BODY_RUNE -> (Math.floor((playerLevel / 46.0)) + 1)
        DefaultRune.COSMIC_RUNE -> (Math.floor((playerLevel / 59.0)) + 1)
        DefaultRune.CHAOS_RUNE -> (Math.floor((playerLevel / 74.0)) + 1)
        DefaultRune.NATURE_RUNE -> (Math.floor((playerLevel / 91.0)) + 1)
        DefaultRune.LAW_RUNE -> 1.0
        DefaultRune.DEATH_RUNE -> 1.0
    }
}
